File: app/src/main/java/com/example/david/lists/di/data/RepositoryModule.java
package com.example.david.lists.di.data;

import com.example.david.lists.data.remote.IRemoteRepository;
import com.example.david.lists.data.repository.IRepository;
import com.example.david.lists.data.repository.RepositoryImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
final class RepositoryModule {
    @Singleton
    @Provides
    IRepository repository(IRemoteRepository remoteRepository) {
        return new RepositoryImpl(remoteRepository);
    }
}

File: app/src/main/java/com/example/david/lists/common/buildlogic/RepositoryModule.kt
package com.example.david.lists.common.buildlogic

import android.app.Application
import com.example.androiddata.repository.buildlogic.RepositoryServiceLocator
import com.example.androiddata.repository.buildlogic.UserRepositoryServiceLocator
import com.example.domain.repository.IRepositoryContract
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class RepositoryModule {
    @Singleton
    @Provides
    fun userRepo(application: Application): IRepositoryContract.UserRepository {
        return UserRepositoryServiceLocator(application, application.packageName).userRepository()
    }

    @Singleton
    @Provides
    fun repo(userRepo: IRepositoryContract.UserRepository): IRepositoryContract.Repository {
        return RepositoryServiceLocator(userRepo).repository()
    }
}